src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -0,0 +1,573 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <boost/iterator/counting_iterator.hpp>
+
+#include "include/intarith.h"
+#include "crimson/os/seastore/journal/circular_bounded_journal.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore {
+
+std::ostream &operator<<(std::ostream &out, const CBJournal::cbj_header_t &header)
+{
+  return out << "cbj_header_t(magin=" << header.magic
+	     << ", uuid=" << header.uuid
+	     << ", block_size=" << header.block_size
+	     << ", size=" << header.size
+	     << ", used_size=" << header.used_size
+	     << ", error=" << header.error
+	     << ", start_offset=" << header.start_offset
+	     << ", applied_to="<< header.applied_to
+	     << ", last_committed_record_base="<< header.last_committed_record_base
+	     << ", written_to=" << header.written_to
+	     << ", flsg=" << header.flag
+	     << ", csum_type=" << header.csum_type
+	     << ", csum=" << header.csum
+	     << ", start=" << header.start
+	     << ", end=" << header.end
+             << ")";
+}
+
+
+CBJournal::CBJournal(NVMeBlockDevice* device, const std::string path)
+  : device(device), path(path) {}
+
+CBJournal::mkfs_ret
+CBJournal::mkfs(mkfs_config_t& config)
+{
+  return _open_device(path
+  ).safe_then([this, &config]() {
+    rbm_abs_addr start_addr = convert_paddr_to_abs_addr(
+      config.start);
+    assert(config.block_size == device->get_block_size());
+    return read_super(start_addr
+    ).safe_then([](auto super) {
+      return mkfs_ertr::now();
```

Comment:
It's not clear to me what's going on here.  Looks like if read_super() suceeds, we return success?  Presumably this is to allow mkfs() to be idempotent?  Should we be checking the returned super against the config uuid?  Certainly, the semantics of this method need to be explained in a comment on the method in the header.